platforms/core-runtime/base-services/src/main/java/org/gradle/internal/service/ServiceScopeValidator.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.service;
+
+import org.gradle.api.NonNullApi;
+import org.gradle.internal.service.scopes.Scope;
+import org.gradle.internal.service.scopes.ServiceScope;
+
+import javax.annotation.Nullable;
+import java.lang.annotation.Annotation;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Checks that services are being declared in the correct scope.
+ * <p>
+ * Only services that are annotated with {@link ServiceScope} are validated.
+ */
+@NonNullApi
+class ServiceScopeValidator implements AnnotatedServiceLifecycleHandler {
+
+    private static final List<Class<? extends Annotation>> SCOPE_ANNOTATIONS = Collections.<Class<? extends Annotation>>singletonList(ServiceScope.class);
```

Comment:
🤔 This could be stored as a single scope and turned into an (immutable) collection only upon `getAnnotations()` being called.

Reply:
I don't think the performance difference would actually be meaningful, but I haven't measured or even looked very closely at how many times `getAnnotations()` would actually be called.

I was a approaching it purely from a readability standpoint.  This is a validator for a single scope.  It's a little unexpected to see it contains a collection of multiple scope annotations.